onpremise/onpremise.go

```diff
@@ -0,0 +1,152 @@
+package onpremise
+
+import (
+	"fmt"
+	"github.com/51Degrees/device-detection-go/v4/dd"
+	"os"
+)
+
+type Engine struct {
+	logger                        logWrapper
+	dataFile                      string
+	licenceKey                    string
+	createTempDataCopy            bool
+	dataUpdateOnStartEnabled      bool
+	dataFileUrl                   string
+	dataFilePullEveryMs           int
+	isScheduledFilePullingEnabled bool
+	loggerEnabled                 bool
+	manager                       *dd.ResourceManager
+	config                        *dd.ConfigHash
+	totalFilePulls                int
+	rdySignal                     chan struct{}
+	fileSynced                    bool
+	isManagerInitialized          bool
+}
+
+// NewResultsHash creates a new ResultsHash object
+// uaCapacity is the initial capacity of the User-Agent cache
+// overridesCapacity is the initial capacity of the overrides cache
+func (p *Engine) NewResultsHash(uaCapacity uint32, overridesCapacity uint32) *dd.ResultsHash {
+	return dd.NewResultsHash(p.manager, uaCapacity, overridesCapacity)
+}
+
+// run starts the pipeline
+func (p *Engine) run() error {
+	if len(p.dataFile) > 0 {
+		err := p.initializeManager()
+		if err != nil {
+			return err
+		}
+	}
+
+	if p.isScheduledFilePullingEnabled {
+		go p.scheduleFilePulling()
+	}
+
+	return nil
+}
+
+func (p *Engine) initializeManager() error {
+	dataFile, err := dd.GetFilePath(".", []string{p.dataFile})
+	if err != nil {
+		return fmt.Errorf("failed to get file path: %w", err)
+	}
+	p.dataFile = dataFile
+
+	err = dd.InitManagerFromFile(p.manager, *p.config, "", p.dataFile)
+	if err != nil {
+		return fmt.Errorf("failed to init manager from file: %w", err)
+	}
+	p.isManagerInitialized = true
+
+	return nil
+}
+
+func (p *Engine) createDatafileIfNotExists() error {
+	_, err := os.Stat(p.dataFile)
+	if err != nil {
+		if len(p.dataFile) == 0 {
+			p.dataFile = "data.hash"
+		}
+
+		_, err = os.Create(p.dataFile)
+		if err != nil {
+			return fmt.Errorf("failed to create data file: %w", err)
+		}
+	}
+
+	return nil
+}
+
+type PipelineOptions func(cfg *Engine) error
+
+func WithDataFile(path string) PipelineOptions {
```

Comment:
please remove `Pipeline` term from use.  This structure should probably be just EngineConfig?   Also With* pattern such as WithDataFile - is something from .NET.  Not sure this pattern is widely accepted in go - perhaps just a constructor  for EngineConfig - the .New that takes all alternative options?   not sure how this is best implemented.